Failure in ExponentialHistogramPercentilesAggregatorTests.testCustomPercentiles

### CI Link

https://gradle-enterprise.elastic.co/s/pvktqhbmurnds/tests/task/:x-pack:plugin:analytics:test/details/org.elasticsearch.xpack.analytics.aggregations.metrics.ExponentialHistogramPercentilesAggregatorTests/testCustomPercentiles?top-execution=1

### Repro line

./gradlew ":x-pack:plugin:analytics:test" --tests "org.elasticsearch.xpack.analytics.aggregations.metrics.ExponentialHistogramPercentilesAggregatorTests.testCustomPercentiles" -Dtests.seed=7C531DAE615C757 -Dtests.locale=es-SV -Dtests.timezone=SystemV/CST6CDT -Druntime.java=25	

### Does it reproduce?

Yes

### Applicable branches

main

### Failure history

https://gradle-enterprise.elastic.co/scans/tests?search.startTimeMax=1773964799999&search.startTimeMin=1773273600000&search.timeZoneId=Europe%2FLondon&tests.container=org.elasticsearch.xpack.analytics.aggregations.metrics.ExponentialHistogramPercentilesAggregatorTests&tests.test=testCustomPercentiles

### Failure excerpt

```
java.lang.AssertionError: Percentiles should be non-decreasing
        at __randomizedtesting.SeedInfo.seed([7C531DAE615C757:BEBBB6D444D49997]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.elasticsearch.xpack.analytics.aggregations.metrics.ExponentialHistogramPercentilesAggregatorTests.lambda$testCustomPercentiles$17(ExponentialHistogramPercentilesAggregatorTests.java:138)
        at org.elasticsearch.search.aggregations.AggregatorTestCase.testCase(AggregatorTestCase.java:881)
        at org.elasticsearch.xpack.analytics.aggregations.metrics.ExponentialHistogramPercentilesAggregatorTests.testCase(ExponentialHistogramPercentilesAggregatorTests.java:204)
        at org.elasticsearch.xpack.analytics.aggregations.metrics.ExponentialHistogramPercentilesAggregatorTests.testCustomPercentiles(ExponentialHistogramPercentilesAggregatorTests.java:130)
```

--- a/libs/exponential-histogram/src/main/java/org/elasticsearch/exponentialhistogram/ExponentialHistogramQuantile.java
+++ b/libs/exponential-histogram/src/main/java/org/elasticsearch/exponentialhistogram/ExponentialHistogramQuantile.java
@@ -70,7 +70,8 @@ public class ExponentialHistogramQuantile {
         if (lowerRank == upperRank) {
             result = values.valueAtRank();
         } else {
-            result = values.valueAtPreviousRank() * (1 - upperFactor) + values.valueAtRank() * upperFactor;
+            double delta = values.valueAtRank() - values.valueAtPreviousRank();
+            result = values.valueAtPreviousRank() + delta * upperFactor;
         }
         return removeNegativeZero(result);
     }

--- a/libs/exponential-histogram/src/test/java/org/elasticsearch/exponentialhistogram/QuantileAccuracyTests.java
+++ b/libs/exponential-histogram/src/test/java/org/elasticsearch/exponentialhistogram/QuantileAccuracyTests.java
@@ -296,7 +296,6 @@ public class QuantileAccuracyTests extends ExponentialHistogramTestCase {
             }
             double exactValue = values[lowerRank] * (1 - upperFactor) + values[upperRank] * upperFactor;
 
-
             // Skip comparison if exact value is close to zero to avoid false-positives due to numerical imprecision
             if (Math.abs(exactValue) < 1e-100) {
                 continue;

--- a/libs/exponential-histogram/src/test/java/org/elasticsearch/exponentialhistogram/QuantileAccuracyTests.java
+++ b/libs/exponential-histogram/src/test/java/org/elasticsearch/exponentialhistogram/QuantileAccuracyTests.java
@@ -42,6 +42,7 @@ import static org.elasticsearch.exponentialhistogram.ExponentialHistogram.MIN_IN
 import static org.elasticsearch.exponentialhistogram.ExponentialScaleUtils.computeIndex;
 import static org.hamcrest.Matchers.closeTo;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 import static org.hamcrest.Matchers.lessThan;
 import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.hamcrest.Matchers.notANumber;
@@ -275,12 +276,18 @@ public class QuantileAccuracyTests extends ExponentialHistogramTestCase {
         double maxError = 0;
         double allowedError = getMaximumRelativeError(values, bucketCount);
 
+        double lastQuantile = Double.NEGATIVE_INFINITY;
         // Compare histogram quantiles with exact quantiles
         for (double q : QUANTILES_TO_TEST) {
             double percentileRank = q * (values.length - 1);
             int lowerRank = (int) Math.floor(percentileRank);
             int upperRank = (int) Math.ceil(percentileRank);
             double upperFactor = percentileRank - lowerRank;
+
+            double histoValue = ExponentialHistogramQuantile.getQuantile(histogram, q);
+            // values should be monotonically increasing
+            assertThat(histoValue, greaterThanOrEqualTo(lastQuantile));
+            lastQuantile = histoValue;
 
             if (values[lowerRank] < 0 && values[upperRank] > 0) {
                 // the percentile lies directly between a sign change and we interpolate linearly in-between
@@ -289,7 +296,6 @@ public class QuantileAccuracyTests extends ExponentialHistogramTestCase {
             }
             double exactValue = values[lowerRank] * (1 - upperFactor) + values[upperRank] * upperFactor;
 
-            double histoValue = ExponentialHistogramQuantile.getQuantile(histogram, q);
 
             // Skip comparison if exact value is close to zero to avoid false-positives due to numerical imprecision
             if (Math.abs(exactValue) < 1e-100) {

